contracts/FlowTransactionScheduler.cdc

```diff
@@ -0,0 +1,1440 @@
+                            error: "Invalid execution effort: \(executionEffort) is greater than the priority's available effort for the requested timestamp."
+                        )
+            }
+
+            if priority == Priority.Low {
+                return EstimatedScheduledTransaction(
+                            flowFee: fee,
+                            timestamp: scheduledTimestamp,
+                            error: "Invalid Priority: Cannot estimate for Low Priority transactions. They will be included in the first block with available space after their requested timestamp."
+                        )
+            }
+
+            return EstimatedScheduledTransaction(flowFee: fee, timestamp: scheduledTimestamp, error: nil)
+        }
+
+        /// calculateScheduledTimestamp calculates the timestamp at which a transaction 
+        /// can be scheduled. It takes into account the priority of the transaction and 
+        /// the execution effort.
+        /// - If the transaction is high priority, it returns the timestamp if there is enough 
+        ///    space or nil if there is no space left.
+        /// - If the transaction is medium or low priority and there is space left in the requested timestamp,
+        ///   it returns the requested timestamp. If there is not enough space, it finds the next timestamp with space.
+        ///
+        /// @param timestamp: The requested timestamp for the transaction
+        /// @param priority: The priority of the transaction
+        /// @param executionEffort: The execution effort of the transaction
+        /// @return UFix64?: The timestamp at which the transaction can be scheduled, or nil if there is no space left for a high priority transaction
+        access(contract) view fun calculateScheduledTimestamp(
+            timestamp: UFix64, 
+            priority: Priority, 
+            executionEffort: UInt64
+        ): UFix64? {
+
+            let used = self.slotUsedEffort[timestamp]
+            // if nothing is scheduled at this timestamp, we can schedule at provided timestamp
+            if used == nil { 
+                return timestamp
+            }
+            
+            let available = self.getSlotAvailableEffort(timestamp: timestamp, priority: priority)
+            // if theres enough space, we can tentatively schedule at provided timestamp
+            if executionEffort <= available {
+                return timestamp
+            } else if priority == Priority.High {
+                // high priority demands scheduling at exact timestamp or failing
+                return nil
+            }
+
+            // if there is no space left for medium or low priority we search for next available timestamp
+            // todo: check how big the callstack can grow and if we should avoid recursion
+            // todo: we should refactor this into loops, because we could need to recurse 100s of times
+            return self.calculateScheduledTimestamp(
```

Comment:
We should convert this into a while loop. The next available timestamp can be far in the future, and this might cause a stack overflow.

Reply:
Issue created here: https://github.com/onflow/flow-core-contracts/issues/538